Reorganization of the repo

W95Psp opened the pull request:
This PR reorganizes and renames the different components of this project:
- `thir-export` is now `circus-frontend`, in the folder `rust-frontend`;
-  `thir-elab` is now`circus-engine`, in the folder `engine`;
-  the directory `cli` contains the Rust crate for the binary of the circus subcommand;
-  the backends live in directories `engine/backends/BACKEND` (with `BACKEND` being `fstar` or `coq` currently)

franziskuskiefer reviewed (approved):
Thanks, this looks much nicer.
There are things (like the one I commented on) that we want to do as well. But not here.
The structure looks good to me. We'll have to talk about the exact names (also to make them publishable on crates.io). But it's a follow-up as welll.

W95Psp commented on cli/Cargo.toml:
```diff
@@ -6,20 +6,20 @@ edition = "2021"
 path = "src/cargo_circus.rs"
 name = "cargo-circus"
 
 [[bin]]
```
I absolutely agree, I also think it's overkill currently, I'll fix that

karthikbhargavan commented:
Any reason why `backends` live inside `engine`? Would be nice if we could pull them out so that anyone can build a backend based on engine, no?

W95Psp commented:
I wanted initially to have a `backend` folder at root, but I was struggling at making opam happy with it, basically... :(

W95Psp commented:
I've added a `diagnostic` crate for hosting errors, cleaned up the JSON schema export binary in Rust, cleaned up the CLI, removed one binary in favor of a subcommand in `circus`

I think we can merge this!
